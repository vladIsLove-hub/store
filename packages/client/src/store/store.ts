import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';

import rootReducer from './reducers/reducer';

const store = createStore(rootReducer, applyMiddleware(thunk));

export default store;

export type RootState = ReturnType<typeof store.getState>;
