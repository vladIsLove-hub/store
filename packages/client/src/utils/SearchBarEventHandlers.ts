import { FocusEvent } from 'react';

const classLists: string[] = ['search-panel__focused', 'search-panel__scale'];

export const focusHandler = (e: FocusEvent<HTMLInputElement>) => {
  e.target.closest('.search-panel')?.classList.add(...classLists);
};

export const blurHandler = (e: FocusEvent<HTMLInputElement>) => {
  e.target.closest('.search-panel')?.classList.remove(...classLists);
};
