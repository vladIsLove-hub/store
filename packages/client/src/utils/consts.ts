export const booksApiPath = '/api/books';
