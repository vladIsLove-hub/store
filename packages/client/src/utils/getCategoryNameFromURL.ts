const getCategoryNameFromURL = (): string | null => {
  const locationPathname = location.pathname;
  const pathnames = locationPathname.split('/');
  const categoryName: string = pathnames[2];
  const page: string = pathnames[3];
  if (locationPathname === `/categories/${categoryName}/${page}` && categoryName) {
    return categoryName;
  } else {
    return null;
  }
};

export default getCategoryNameFromURL;
