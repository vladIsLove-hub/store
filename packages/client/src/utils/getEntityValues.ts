import {
  PlainTextValue,
  SingleEntityValue,
  MultipleEntityValue,
} from '../components/book-view-item/details/details.types';

type GetEntityReturnedValues = {
  entityValue?: string;
  entity?: SingleEntityValue;
  entityList?: MultipleEntityValue;
  entityNames?: string;
};

const getEntityValues = (value: PlainTextValue | SingleEntityValue | MultipleEntityValue): GetEntityReturnedValues => {
  if (typeof value === 'string' || typeof value === 'number') {
    return {
      entityValue: value.toString(),
    };
  }

  if (Array.isArray(value)) {
    const nameList = value
      .map(entity => {
        return entity.name;
      })
      .join(', ');
    return {
      entityList: value,
      entityNames: nameList,
    };
  }

  return {
    entity: value,
  };
};

export default getEntityValues;
