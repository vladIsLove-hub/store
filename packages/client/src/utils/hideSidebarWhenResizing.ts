export const hidenSidebarWhenResizing = (handleClose: () => void): void => {
  window.addEventListener('resize', (e: UIEvent) => {
    const w = e.target as Window;
    const resolutionForTablets = 990;
    if (w.innerWidth > resolutionForTablets) {
      handleClose();
    }
  });
};

export default hidenSidebarWhenResizing;
