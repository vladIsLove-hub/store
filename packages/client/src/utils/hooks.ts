import { useSelector } from 'react-redux';

import { RootState } from '../store/store';

export const useCustomSelector = () => {
  return useSelector((state: RootState) => {
    return state;
  });
};
