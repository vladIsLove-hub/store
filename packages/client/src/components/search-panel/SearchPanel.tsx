import React, { FC, useState } from 'react';
import { useDebouncedCallback } from 'use-debounce/lib';
import { useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';

import SearchIcon from '../../icons/SearchIcon.svg';
import { focusHandler, blurHandler } from '../../utils/SearchBarEventHandlers';
import { Book } from '../booklist/BookListTypes';
import { searchChangeHandler, setSearchData } from '../../store/action-creators/search.actions';

import SearchDropdown from './search-dropdown/SearchDropdown';

const SearchPanel: FC = () => {
  const history = useHistory();
  const dispatch = useDispatch();
  const [search, setSearch] = useState<string>('');
  const [books, setBooks] = useState<Book[]>([]);
  const requestDelay = 300;
  const debounced = useDebouncedCallback(async e => {
    await searchChangeHandler<Book[]>(e, setBooks);
  }, requestDelay);

  const clickHandler = (): void => {
    setSearchData(dispatch, search);
    setSearch('');
    setBooks([]);
    history.push('/');
  };

  return (
    <div className="search-panel">
      <img src={SearchIcon} alt="" className="search-panel__icon" />
      <input
        value={search}
        onFocus={focusHandler}
        onBlur={blurHandler}
        type="search"
        placeholder="Enter book name or author-name..."
        className="search-panel__input"
        onChange={e => {
          setSearch(e.target.value);
          debounced(e);
        }}
      />
      <button
        disabled={Boolean(!search)}
        onClick={clickHandler}
        className={`search-panel__btn ${search ? '' : 'search-panel__btn-disabled'}`}
      >
        Search
      </button>
      <SearchDropdown books={books} setBooks={setBooks} setSearch={setSearch} />
    </div>
  );
};

export default SearchPanel;
