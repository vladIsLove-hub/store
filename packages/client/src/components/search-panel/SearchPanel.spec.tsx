import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Provider } from 'react-redux';

import { focusHandler, blurHandler } from '../../utils/SearchBarEventHandlers';
import store from '../../store/store';

import SearchPanel from './SearchPanel';

describe('SearchPanel', () => {
  it('should render correctly', () => {
    render(
      <Provider store={store}>
        <SearchPanel />
      </Provider>,
    );
    expect(screen.getByText('Search')).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/Enter book name.../i)).toBeInTheDocument();
  });

  it('focusHandler defined', () => {
    expect(focusHandler).toBeDefined();
  });

  it('blurHandler defined', () => {
    expect(blurHandler).toBeDefined();
  });

  it('input styles ixists', () => {
    render(
      <Provider store={store}>
        <SearchPanel />
      </Provider>,
    );
    expect(screen.getByPlaceholderText(/Enter book name.../i)).toHaveClass('search-panel__input');
  });

  it('typing in search input', () => {
    render(
      <Provider store={store}>
        <SearchPanel />
      </Provider>,
    );
    expect(screen.queryByDisplayValue(/Some/)).toBeNull();
    userEvent.type(screen.getByRole('searchbox'), 'Some');
    expect(screen.queryByDisplayValue(/Some/));
  });

  it('Search snapshot', () => {
    const search = render(
      <Provider store={store}>
        <SearchPanel />
      </Provider>,
    );
    expect(search).toMatchSnapshot();
  });
});
