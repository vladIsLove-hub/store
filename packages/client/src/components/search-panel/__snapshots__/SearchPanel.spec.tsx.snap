// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPanel Search snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="search-panel"
      >
        <img
          alt=""
          class="search-panel__icon"
          src="SearchIcon.svg"
        />
        <input
          class="search-panel__input"
          placeholder="Enter book name or author-name..."
          type="search"
          value=""
        />
        <button
          class="search-panel__btn search-panel__btn-disabled"
          disabled=""
        >
          Search
        </button>
        <div
          class="search-dropdown "
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="search-panel"
    >
      <img
        alt=""
        class="search-panel__icon"
        src="SearchIcon.svg"
      />
      <input
        class="search-panel__input"
        placeholder="Enter book name or author-name..."
        type="search"
        value=""
      />
      <button
        class="search-panel__btn search-panel__btn-disabled"
        disabled=""
      >
        Search
      </button>
      <div
        class="search-dropdown "
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
